joomla/joomla-cms
# Fix error 404 in modern routing for links without own menu item (NomenuRules)

### Summary of Changes

When there is no menu item for category/categories/article view in com_content and option `Remove IDs` is enabled then a link like:
- `/index.php/en/component/content/article/article-en-gb?catid=8` returns error 404

In this PR  I rewrote `NomenuRules` class to `build` and `parse` links:
- `index.php/en/component/content/article/category-en-gb/article-en-gb` (option `Remove IDs` is enabled)
- `index.php/en/component/content/article/3-category-en-gb/2-article-en-gb`


### Testing Instructions
Disable all menu items for `com_content` `categories/category/article` views.
Go to frontend `index.php/component/content/categories` and check links to category and article items.

### Expected result
Links work, no 404 when option `Remove Ids` is enabled/disabled.

### Actual result
When option `Remove Ids` is enabled then links to articles return error 404.
